### DISCUSSION TITLE
'stdlib.h' not found error while compiling with syclcc

### DISCUSSION THREAD

**Discussion by Surya99941** [AdaptiveCpp/AdaptiveCpp#1015] [Q&A]
Arch Linux, installation through AUR, 
packages installed:
- hipsycl-cpu
- hipsycl-cuda

here is c++ code i ran:
```
#include <CL/sycl.hpp>

#include <vector>
#include <iostream>

using T = float;

int main() {
	std::vector<T> a = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
	std::vector<T> b = { 9, 8, 7, 6, 5, 4, 3, 2, 1,  0 };
	std::vector<T> c(10);

	cl::sycl::queue q;

	std::cout
		<< "Running on "
		<< q.get_device().get_info<cl::sycl::info::device::name>()
		<< std::endl;

	cl::sycl::buffer<T> A(a.data(), a.size());
	cl::sycl::buffer<T> B(b.data(), b.size());

	{
		cl::sycl::buffer<T> C(c.data(), c.size());

		q.submit([&](cl::sycl::handler &cgh) {
			auto ka = A.get_access<cl::sycl::access::mode::read>(cgh);
			auto kb = B.get_access<cl::sycl::access::mode::read>(cgh);
			auto kc = C.get_access<cl::sycl::access::mode::write>(cgh);

			cgh.parallel_for<class vector_add>(
				cl::sycl::range<1>{10},
				[=](cl::sycl::id<1> index) {
					kc[index] = ka[index] + kb[index];
				}
			);
		});
	}

	std::cout << std::endl << "Result:" << std::endl;

	for (auto elem : c) {
		std::cout << elem << " ";
	}
	std::cout << std::endl;
}
```

I compiled the code using following command:
`syclcc ./main.cpp -o ./main -fsycl -I/usr/include/ --hipsycl-targets="omp"`

And the error i got is
```
syclcc warning: No optimization flag was given, optimizations are disabled by default. Performance may be degraded. Compile with e.g. -O2/-O3 to enable optimizations.
In file included from ./main.cpp:54:
In file included from /usr/bin/../include/CL/sycl.hpp:31:
In file included from /usr/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:50:
In file included from /usr/bin/../include/hipSYCL/glue/persistent_runtime.hpp:32:
In file included from /usr/bin/../include/hipSYCL/runtime/application.hpp:33:
In file included from /usr/bin/../include/hipSYCL/runtime/backend.hpp:32:
In file included from /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/12.2.1/../../../../include/c++/12.2.1/string:53:
In file included from /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/12.2.1/../../../../include/c++/12.2.1/bits/basic_string.h:3960:
In file included from /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/12.2.1/../../../../include/c++/12.2.1/ext/string_conversions.h:41:
/usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/12.2.1/../../../../include/c++/12.2.1/cstdlib:75:15: fatal error: 'stdlib.h' file not found
#include_next <stdlib.h>
              ^~~~~~~~~~
1 error generated.
```

**Comment by illuhad** [comment]
Hi!

a couple of pointers:
* The `hipsycl-*` AUR packages are maintained by third parties. I don't know how these packages work these days or if they follow best practices. From a quick glance it seems that these packages are *not* set up to support multiple backends simultaneously.
* Never use `-fsycl`. `-fsycl` is a flag that has been upstreamed from DPC++, i.e. a different SYCL implementations. Activating two SYCL implementations simultaneously is obviously not a good idea :-)
* The error you see seems to indicate that your basic C++ standard library does not work correctly. Can you confirm whether you can correctly translate a basic C++ source file using clang and gcc that includes `<cstdlib>`?

**Reply by Surya99941** [reply]
installed from source and now its working perfectly fine, 
A liitle problem is how to force compile using nvcc compiler instead of clang compiler for cuda

**Reply by illuhad** [reply]
There is no way to use nvcc because nvcc does not support automatic kernel outlining, which is a prerequisite for SYCL.

We do support using `nvc++` from NVIDIA's HPC SDK though. The `cuda` target uses clang, while the `cuda-nvcxx` target uses nvc++. That said, for most cases the recommendation is to use clang:
* clang compiles much, much faster than nvc++ (or nvcc)
* The NVIDIA compilers are far less stable, and nvc++ is prone to compiler crashes
* Performance between clang and nvc++ is usually very close, with clang usually outperforming nvc++ more often than vice versa (we are talking about ~10% differences here, so nothing huge).

The main use cases for preferring nvc++ over clang are:
* Being on an HPC cluster where the NVIDIA HPC SDK is already preinstalled, while clang/LLVM must first be manually compiled from source. In that case, it is much simpler to install hipSYCL when using nvc++ and thereby getting around the LLVM dependency
* Having access to NVIDIA hardware that is so brand-new and/or secret that clang does not yet support it.


Apart from very specific use cases there really is little reason to use the proprietary compilers.
